azure-iot-device/azure/iot/device/common/pipeline/pipeline_ops_base.py

```diff
@@ -241,10 +267,15 @@ def __init__(self, request_type, method, resource_location, request_body, reques
         :param Function callback: The function that gets called when this operation is complete or has
           failed.  The callback function must accept A PipelineOperation object which indicates
           the specific operation which has completed or failed.
+        :param str registration_id: The registration ID is used to uniquely identify a device in the Device Provisioning Service.
+        :param str operation_id: The id of the operation as returned by the initial provisioning request.
         """
         super(RequestOperation, self).__init__(callback=callback)
         self.method = method
         self.resource_location = resource_location
         self.request_type = request_type
         self.request_body = request_body
         self.request_id = request_id
+        # TODO Question : If not here we have to define new op DPSRequestOperation
+        self.registration_id = registration_id
+        self.operation_id = operation_id
```

Comment:
I can see why you need these here, and I can see why they're useful, but they don't belong in this object as you have them today, but I don't think you need to add a DPSRequestOperation object either. I would suggest:

1. registration_id does not affect the topic.  It only affects the payload.  I would rather you create the DeviceRegistrationPayload object in pipeline_starges_provisioing.py instead of pipeline_stages_provisioning_mqtt.py.  That way, you don't have to pass registration_id down in the op.

2. operation_id does affect the topic, and it's like a query parameter on a URL, so I would add query_parameters to RequestOperation.  In pipeline_stages_provisioning, you could set op.query_parameters={"operationId", operation_id} before you pass it down, and in pipeline_stages_provisioning_mqtt.py, you could get the operation_id out of this dict when you build the topic
This way, RequestOperation still carries the info you need, but it's not stored in DPS-specific attributes.

Reply:
good idea. i also need query params in the requestandresponseoperation as well.

---
In reply to: [345896656](https://github.com/Azure/azure-iot-sdk-python/pull/391#discussion_r345896656) [](ancestors = 345896656)